torchdynamo/variables/torch.py

```diff
@@ -148,6 +149,56 @@ def call_function(
         elif not config.dynamic_shapes and self.is_dynamic_shapes(args, kwargs):
             unimplemented(f"dynamic shapes: {self.value.__name__}")
         else:
+            # print('start of torch.py else branch')
+            # TODO: this is just a hack to get an e2e example working,
+            # needs a lot of cleanup before review
+            if self.value == torch.nn.functional.sigmoid:
+                # this extracts the tensor under test
+                # TODO: is this always there and can we depend on it?
+                tensor_val = args[0].proxy.node.meta['example_value']
+                from torch.overrides import has_torch_function_unary
+                has_tf = has_torch_function_unary(tensor_val)
+                if has_tf:
+                    from torchdynamo.source import LocalSource
+                    from torchdynamo.variables.builder import VariableBuilder, TupleVariable
+                    # note: does this actually need to be a UserFunctionVariable?
+                    # note: the __func__ attribute call is to ensure the function
+                    # does not have `self`
+                    func = tensor_val.__torch_function__.__func__
+
+                    # For now, we inline the object instance's __torch_function__ directly
+                    # In the future, it might be worth considering inlining
+                    # the `torch.overrides.handle_torch_function` method instead.
+
+                    # first, get the source
+                    # TODO: is this the right type of source to use?
+                    source = LocalSource('__torch_function__')
```

Comment:
This line is the root cause of your `NameError: name '__torch_function__' is not defined` error.

`LocalSource('__torch_function__')` tells the guard that it can get this value from the local `'__torch_function__'`, which doesn't exist and triggers the error.

The "minimal" fix would be:
```suggestion
                    assert args[0].source
                    source = AttrSource(AttrSource(args[0].source, '__torch_function__'), '__func__')
```

Which, will cause the guard to read `tensor.__torch_function__.__func__`.


----

You might want to try the above just to convince yourself it works.

However, I think a better approach would be:

1) Check for the existence of `__torch_function__` in `variable_builder.py`
2) Map tensors with `__torch_function__` to a new `TensorSubclassVariable` instead of `TensorVariable`. `TensorSubclassVariable` could contain a `TensorVariable` as a property.  We do this because we want to eliminate the tensor subclass from the generated code, so don't want the default tensor handling.
3) When we do operations on a `TensorSubclassVariable`, just inline calls to `__torch_function__` similar to what you do here.